tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,260 @@
+          lambda chain_state: chain_state.shape,
+          initial_chain_state)
+      self._parameters['dtype'] = tf.nest.map_structure(
+          lambda chain_state: chain_state.dtype,
+          initial_chain_state)
+      if self.event_ndims is None:
+        self._parameters['event_ndims'] = tf.nest.map_structure(
+            ps.rank, initial_chain_state)
+      elif not nest.is_sequence(self.event_ndims):
+        self._parameters['event_ndims'] = tf.nest.map_structure(
+            lambda _: self.event_ndims,
+            initial_chain_state,
+        )
+      self.strms = nest.map_structure_up_to(
+          # only parameter guaranteed to give desired shallow structure
+          self._parameters['dtype'],
+          sample_stats.RunningCovariance,
+          self.shape, self.event_ndims, self.dtype,
+          check_types=False,
+      )
+      return tf.nest.map_structure(
+          lambda strm: strm.initialize(),
+          self.strms)
+
+
+  def one_step(
+      self,
+      sample,
+      current_reducer_state,
+      previous_kernel_results=None,
+      axis=None):
+    """Update the `current_reducer_state` with a new sample.
+
+    Chunking semantics are similar to those of batching and are specified by the
+    `axis` parameter. If chunking is enabled (axis is not `None`), all elements
+    along the specified `axis` will be treated as separate samples. If a
+    single scalar value is provided for a non-scalar sample structure, that
+    value will be used for all samples in the latent. If not, an identical
+    structure must be provided.
+
+    Args:
+      sample: A (possibly nested) structure of incoming sample(s) with shape
+        and dtype compatible with those used to initialize the
+        `current_reducer_state`.
+      current_reducer_state: A (possibly nested) structure of
+        `RunningCovarianceState`s representing the current state of the running
+        covariance.
+      previous_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made in a related
+        `TransitionKernel`. For streaming covariance, this argument has no
+        influence on computation; hence, it is `None` by default. However, this
+        argument is still accepted as it will be supplied by the Streaming MCMC
+        framework.
+      axis: If chunking is desired, this is a (possibly nested) structure of
+        integers that specifies the axis with chunked samples. For individual
+        samples, set this to `None`. By default, samples are not chunked
+        (`axis` is None).
+
+    Returns:
+      new_state: `RunningCovarianceState` with updated running statistics and
+        identical structure to the `current_reducer_state`.
+    """
+    with tf.name_scope(
+        mcmc_util.make_name(self.name, 'covariance_reducer', 'one_step')):
+      if not nest.is_sequence(axis):
+        axis = tf.nest.map_structure(
```

Comment:
Have you considered using `tfp.python.internal.nest_util.broadcast_structure` here (and elsewhere)?

Reply:
Didn't even know this existed. Helps a lot!